### PR TITLE
feat: read data from write cache

### PR DESCRIPTION
**Pull request by QuenKar**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
part of #2965

changes:
- move fetch ranges function to `helper.rs`.
- add `read_ranges` function for `FileCache`.
- support to try reading from write cache first, if failed, read from remote object store directly.
- tiny refactor `usize` to `u64`

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
#2965

### CODE REVIEW

File: src/mito2/src/sst/parquet/row_group.rs [hunk 1]
```diff
@@ -166,19 +184,28 @@ impl<'a> InMemoryRowGroup<'a> {
...
-                    .into_iter();
+                if let Some(data) = self.fetch_ranges_from_write_cache(key, &fetch_ranges).await {
+                    data.into_iter()
+                } else {
+                    // Fetch data from object store directly.
+                    fetch_byte_ranges(self.file_path, self.object_store.clone(), &fetch_ranges)
+                        .await
+                        .map_err(|e| ParquetError::External(Box::new(e)))?
+                        .into_iter()
+                };
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/row_group.rs:208]: Consider wrapping a function or struct to do this and eliminating some duplicated codes.

File: src/mito2/src/sst/parquet/row_group.rs [hunk 2]
```diff
@@ -115,21 +117,37 @@ impl<'a> InMemoryRowGroup<'a> {
...
                         _ => (),
                     }
 
-                    ranges.extend(selection.scan_ranges(&page_locations[idx]));
-                    page_start_offsets.push(ranges.iter().map(|range| range.start).collect());
+                    ranges.extend(
+                        selection
+                            .scan_ranges(&page_locations[idx])
+                            .iter()
+                            .map(|r| r.start as u64..r.end as u64),
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/row_group.rs:129]:
```suggestion
                            .map(|range| range.start as u64..range.end as u64),
```

File: src/mito2/src/sst/parquet.rs [hunk 3]
```diff
@@ -15,7 +15,7 @@
 //! SST in parquet format.
 
 mod format;
-mod helper;
+pub mod helper;
```
Comment (evenyag) [inline src/mito2/src/sst/parquet.rs:18]: `pub(crate)` should be enough. Do not expose too many details unless it is necessary.

File: src/mito2/src/cache/file_cache.rs [hunk 4]
```diff
@@ -129,6 +132,36 @@ impl FileCache {
...
+            CACHE_MISS.with_label_values(&[FILE_TYPE]).inc();
+            return None;
+        }
+
+        let file_path = self.cache_file_path(key);
+        // In most cases, it will use blocking read,
+        // because FileCache is normally based on local file system, which supports blocking read.
+        let bytes_result = fetch_byte_ranges(&file_path, self.local_store.clone(), ranges).await;
+
+        match bytes_result {
```
Comment (evenyag) [inline src/mito2/src/cache/file_cache.rs:151]:
```suggestion
        let bytes_result = fetch_byte_ranges(&file_path, self.local_store.clone(), ranges).await;
        match bytes_result {
```

File: src/mito2/src/cache/file_cache.rs [hunk 5]
```diff
@@ -129,6 +132,36 @@ impl FileCache {
...
+            Ok(bytes) => {
+                CACHE_HIT.with_label_values(&[FILE_TYPE]).inc();
+                Some(bytes)
+            }
+            Err(_) => {
+                // We removes the file from the index.
+                self.memory_index.remove(&key).await;
+                CACHE_MISS.with_label_values(&[FILE_TYPE]).inc();
```
Comment (evenyag) [inline src/mito2/src/cache/file_cache.rs:159]: We should log the error like `reader()` if it is not `NotFound`.

File: src/mito2/src/sst/parquet/row_group.rs [hunk 6]
```diff
@@ -166,19 +184,28 @@ impl<'a> InMemoryRowGroup<'a> {
...
+            let key = (self.region_id, self.file_id);
             let mut chunk_data =
-                fetch_byte_ranges(self.file_path, self.object_store.clone(), fetch_ranges)
-                    .await?
-                    .into_iter();
+                if let Some(data) = self.fetch_ranges_from_write_cache(key, &fetch_ranges).await {
+                    data.into_iter()
+                } else {
+                    // Fetch data from object store directly.
+                    fetch_byte_ranges(self.file_path, self.object_store.clone(), &fetch_ranges)
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/row_group.rs:204]: Can we add metrics to collect the cost of fetching pages on a cache miss? Maybe we can reuse the `READ_STAGE_ELAPSED` metrics or adding a new one.

**Review by zhongzc — APPROVED** [review]
most LGTM

**Comment by killme2008** [comment]
After we merge this PR, could we remove the cache layer?

**Comment by QuenKar** [comment]
> After we merge this PR, could we remove the cache layer?


couldn't be removed yet, because we doesn't implement cache read file now, if data is not in WriteCache and we still rely on cache layer to cache read data. The Write Cache just caches the write file and upload to remote store.

**Comment by killme2008** [comment]
> > After we merge this PR, could we remove the cache layer?
> 
> couldn't be removed yet, because we doesn't implement cache read file now, if data is not in WriteCache and we still rely on cache layer to cache read data. The Write Cache just caches the write file and upload to remote store.

Got it. I believe our final goal is to merge these two components.

**Review by evenyag — APPROVED** [review]
LGTM.

Could we add a case to test the engine with write cache enabled under [mito2::engine](https://github.com/GreptimeTeam/greptimedb/tree/main/src/mito2/src/engine) mod? It could be done in the next PR.

**Pull request merged**